x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/preview_rules_route.ts

```diff
@@ -184,6 +189,32 @@ export const previewRulesRoute = async (
               tags: [],
               updatedBy: rule.updatedBy,
             })) as TState;
+
+            // Save and reset error and warning logs
+            const currentErrors = {
+              logs: warningsAndErrorsStore
+                .filter((item) => item.newStatus === RuleExecutionStatus.failed)
+                .map((item) => item.message),
+              startedAt: startedAt.toDate().toISOString(),
+            };
+            const currentWarnings = {
+              logs: warningsAndErrorsStore
+                .filter(
+                  (item) =>
+                    item.newStatus === RuleExecutionStatus['partial failure'] ||
+                    item.newStatus === RuleExecutionStatus.warning
+                )
+                .map((item) => item.message),
+              startedAt: startedAt.toDate().toISOString(),
+            };
+            if (currentErrors.logs.length > 0) {
+              errors.push(currentErrors);
+            }
+            if (currentWarnings.logs.length > 0) {
+              warnings.push(currentWarnings);
```

Comment:
Instead of storing the warnings as a separate array where each invocation only has an entry if the invocation generated a warning, it would be nice to return a single array of objects for both errors and warnings where each invocation always has an entry. E.g.
```
[
  {startedAt: 1st date, errors: [], warnings: []}, 
  {startedAt: 2nd date, errors: [your error object here], warnings: []},
  ...
]
```
This way we can (1) easily associated errors and warnings that came from the same rule execution, and (2) we have a place to add more status information for each invocation in the future.